server/src/main/java/nl/inl/blacklab/server/index/IndexManager.java

```diff
@@ -550,20 +564,39 @@ private synchronized void loadUserIndices(String userId) {
     }
 
     /**
-     * Checks all indices to see if their directories are still readable, and
-     * removes them if this is not the case. This can happen when the index is
-     * deleted on-disk while we're running. This feature is explicitly supported.
+     * Starts a monitor to remove references to indices whose physical file was removed
+     * @param directories to monitor
+     * @param pollingIntervalInMs how ofter to monitor the directories
+     * @return the monitor
+     * @throws Exception
      */
-    private synchronized void cleanupRemovedIndices() {
-        List<String> removedIds = new ArrayList<>();
-        for (Index i : indices.values()) {
-            if (!i.getDir().canRead()) {
-                removedIds.add(i.getId());
+    public FileAlterationMonitor startRemovedIndicesMonitor(List<File> directories, long pollingIntervalInMs) throws Exception {
+        logger.info("Installing index removal watcher on: {}", directories);
+        FileAlterationMonitor monitor = new FileAlterationMonitor(pollingIntervalInMs);
+        List<FileAlterationObserver> observers = directories.stream()
+            .map(FileAlterationObserver::new)
+            .collect(Collectors.toList());
+        FileAlterationListenerAdaptor listener = new FileAlterationListenerAdaptor() {
+            @Override
+            public void onDirectoryDelete(File directory) {
+                logger.info("Directory deleted: {}", directory.getAbsolutePath());
+                synchronized (IndexManager.this) {
+                    Optional<Index> indexToDelete = indices.values().stream()
+                        .filter(i -> i.getDir().equals(directory))
+                        .findFirst();
+                    indexToDelete.ifPresent(i -> {
+                        logger.info("Deleting index {}, {}", i.getId(), i.getDir().getAbsolutePath());
+                        indices.remove(i.getId());
```

Comment:
index.close() disappeared here, that should be re-added or there will be some stale references elsewhere in the code

Reply:
great catch! thanks @KCMertens